src/storage-controller/src/instance.rs

```diff
@@ -231,38 +295,263 @@ where
-                    self.active_ingestions.insert(ingestion.id);
+                // First absorb into our state, because this might change
+                // scheduling decisions, which need to be respected just below
+                // when sending commands.
+                self.absorb_ingestions(ingestions.clone());
+
+                for cmd in ingestions.iter() {
+                    for replica in self.active_replicas(&cmd.id) {
+                        replica.send(StorageCommand::RunIngestions(vec![cmd.clone()]));
+                    }
                 }
             }
             StorageCommand::RunSinks(sinks) => {
-                for sink in sinks {
+                for sink in sinks.iter() {
                     self.active_exports.insert(sink.id);
                 }
+                for replica in self.replicas.values_mut() {
+                    replica.send(command.clone());
+                }
             }
-            StorageCommand::AllowCompaction(policies) => {
-                for (id, frontier) in policies {
-                    if frontier.is_empty() {
-                        self.active_ingestions.remove(id);
-                        self.active_exports.remove(id);
+            StorageCommand::AllowCompaction(cmds) => {
+                // First send out commands and then absorb into our state since
+                // absorbing them might remove entries from active_ingestions.
+                for (id, frontier) in cmds.iter() {
+                    for replica in self.active_replicas(id) {
+                        replica.send(StorageCommand::AllowCompaction(vec![(
+                            id.clone(),
+                            frontier.clone(),
+                        )]));
                     }
                 }
+
+                self.absorb_compactions(cmds);
+            }
+            command => {
+                for replica in self.replicas.values_mut() {
+                    replica.send(command.clone());
+                }
             }
-            _ => (),
         }
 
-        // Record the command so that new replicas can be brought up to speed.
-        self.history.push(command.clone());
+        if command.installs_objects() && self.replicas.is_empty() {
+            self.update_paused_statuses();
+        }
+    }
 
-        // Clone the command for each active replica.
-        for replica in self.replicas.values_mut() {
-            replica.send(command.clone());
+    /// Updates internal state based on incoming ingestion commands.
+    ///
+    /// This does _not_ send commands to replicas, we only record the ingestion
+    /// in state and potentially update scheduling decisions.
+    fn absorb_ingestions(&mut self, ingestions: Vec<RunIngestionCommand>) {
+        for ingestion in ingestions {
+            let prefers_single_replica = ingestion
+                .description
+                .desc
+                .connection
+                .prefers_single_replica();
```

Comment:
I see, so this method already exists there. Is there a reason we can't use this instead of de-normalizing it in the `ActiveIngestion` struct?

Reply:
By "already exist", you mean it's added in this PR? 😅 

But yes, see my answer up here: https://github.com/MaterializeInc/materialize/pull/31890/files/9413c5366af22285e88e7c7ad279fcfcf1a20d65#r1999229940